### PR TITLE
Update stream to use `--exit-handshake`

### LINKED ISSUE #8: Implement exit handshaking to avoid dropping final messages
There's a race condition between Erlang processing the exit code and it processing the messages that get sent to it. For example, `fwup` can fail and send an error message up. `fwup` will exit after it does this, but Erlang may handle the process exit first and lose the error message. That's not helpful.

The fix to this is to pass the `--exit-handshake` option to `fwup`. If this is done, then after `fwup` reports `OK` or `ER`, it will sent a CTRL+Z character (this part doesn't matter with framing), and then it will call `read()` and wait for Erlang to send something back to it. That's the handshake. All Erlang needs to do is send a byte (anything works) back to `fwup` after it gets an `OK` or `ER`. Then `fwup` will exit knowing that Erlang safely got the error or success message.

To summarize, the change is:

1. Pass `--exit-handshake` when calling `fwup`
2. Send a character to the port after receiving an `OK` or `ER`

After sending the character, it's ok to finish up completely. I don't think there's a need to wait for `fwup` to exit since you'll know that the exit code is 0 if the last message is `OK` or 1 if `ER`. Also, closing the port will also satisfy the handshake from `fwup`'s point of view.

Another note is that it would be really bad for `fwup` to exit without sending an `OK` or `ER`. That would be a bad bug in fwup if that ever happens, so it's worth adding an ugly log message blaming fwup.

### PR DESCRIPTION
**Pull request by ConnorRigby**
Fixes #8

### CODE REVIEW

**Comment by fhunleth** [comment]
This looks right. This bug is kind of annoying due to it being a race condition. Out of curiosity, does the unit test fail without the fix?

**Comment by ConnorRigby** [comment]
it didn't seem to fail without it. i suspect if i ran it enough times it might tho, being a race condition. I've never seen this race condition before in real usage

File: lib/fwup/stream.ex [hunk 1]
```diff
@@ -52,27 +52,27 @@ defmodule Fwup.Stream do
...
+  def handle_info({port, {:data, <<"OK", code::16, message::binary>>}}, state) do
+    dispatch(state, {:ok, code, message})
+    _ = Port.command(port, "OK")
+    _ = Port.close(port)
     {:stop, :normal, %{state | port: nil}}
   end
 
   def handle_info({_port, {:exit_status, status}}, state) do
-    {:stop, {:exit, status}, %{state | port: nil}}
+    {:stop, {:error, {:unexpected_exit, status}}, %{state | port: nil}}
```
Comment (jjcarstens) [inline lib/fwup/stream.ex]: Won't we still get the exit_status here even if we are manually doing `Port.close` and it exits normal?

Comment (ConnorRigby) [reply lib/fwup/stream.ex]: the clause that does the `Port.close` returns `{:stop, :normal, %{state | port: nil}` so i don't think the exit status message will get handled

**Comment by jjcarstens** [comment]
@coop is running into this. For posterity, these were the logs that led to the conversation with Frank which led to this:

```
00:26:52.104 [info]  [NervesHubLink] Downloading firmware: https://some-firmware-link
00:26:52.180 [error] GenServer Fwup.Stream terminating
** (stop) {:exit, 1}
Last message: {#Port<0.109>, {:exit_status, 1}}
State: %{cm: #PID<0.1686.0>, port: nil}
00:26:52.187 [error] HTTP Streaming Error: {{:exit, 1}, {GenServer, :call, [#PID<0.1778.0>, {:get, "https://some-firmware-link"}, :infinity]}}
00:27:00.211 [info]  [NervesHubLink] Downloading firmware: https://some-firmware-link
00:27:00.228 [error] GenServer Fwup.Stream terminating
** (stop) {:exit, 1}
Last message: {#Port<0.112>, {:exit_status, 1}}
State: %{cm: #PID<0.1686.0>, port: nil}
00:27:00.241 [error] HTTP Streaming Error: {{:exit, 1}, {GenServer, :call, [#PID<0.1785.0>, {:get, "https://some-firmware-link"}, :infinity]}}
```

This is _all_ the log, which is lacking any warning or error that should have been returned from fwup.

**Comment by fhunleth** [comment]
I wish that I knew of a fool proof way of making it happen. It's subtle, though. I think that since firmware updates are overwhelmingly successful and I doubt anyone cares if they miss an "OK" message when they get a success exit.

You might be able to induce it by putting a sleep on the progress message handling - probably a 100ms sleep or something small like that. Obviously, we don't want that sleep in production code, but I'd really like to know that this fix is sufficient.

File: lib/fwup/stream.ex [hunk 2]
```diff
@@ -52,27 +52,27 @@ defmodule Fwup.Stream do
...
     dispatch(state, {:warning, code, message})
     {:noreply, state}
   end
 
-  def handle_info({_port, {:data, <<"OK", code::16, message::binary>>}}, state) do
-    dispatch(state, {:ok, code, message})
-    {:noreply, state}
+  def handle_info({port, {:data, <<"ER", code::16, message::binary>>}}, state) do
+    dispatch(state, {:error, code, message})
+    _ = Port.command(port, "ER")
```
Comment (fhunleth) [inline lib/fwup/stream.ex]: I double-checked the fwup code. I told you wrong, nothing needs to be sent. The port just needs to be closed. You can delete lines 62 and 69.

**Comment by ConnorRigby** [comment]
i introduced a `Process.sleep/1` on the `progress` clause. I think i see the race condition, but not in a way i was expecting.
```elixir
08:18:30.940 [error] GenServer #PID<0.188.0> terminating
** (ArgumentError) argument error
    :erlang.port_close(#Port<0.6>)
    (fwup) lib/fwup/stream.ex:69: Fwup.Stream.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Last message: {#Port<0.6>, {:data, <<79, 75, 0, 0>>}}
State: %{cm: #PID<0.185.0>, port: #Port<0.6>}
```

I believe the port was closed at the same time or just moments before the `Port.close(port)` call.

**Comment by coop** [comment]
With this change I am now seeing fwup errors in the logs:

```
19:11:13.190 [error] GenServer Fwup.Stream terminating
** (stop) "Error decoding public key"
Last message: {#Port<0.102>, {:data, <<69, 82, 0, 0, 69, 114, 114, 111, 114, 32, 100, 101, 99, 111, 100, 105, 110, 103, 32, 112, 117, 98, 108, 105, 99, 32, 107, 101, 121>>}}
State: %{cm: #PID<0.1682.0>, port: nil}
19:11:13.191 [error] FWUP ERROR: Error decoding public key
19:11:13.199 [error] HTTP Streaming Error: {"Error decoding public key", {GenServer, :call, [#PID<0.1766.0>, {:get, "https://s3.amazonaws.com/nerves-hub/firmware/..."}, :infinity]}}
```

Thanks!

**Comment by fhunleth** [comment]
@coop I hope that error turns out to be helpful. Thanks for letting us know that the fix worked!!!

@ConnorRigby I'm approving. I saw that CI is failing. Not sure why, but I'm good with the change. Thanks for fixing it so quickly.

**Comment by ConnorRigby** [comment]
@fhunleth CI failing is due to an unrelated issue. Will update it after i merge this PR

**Pull request merged**